lta/rest_server.py

```diff
@@ -370,6 +368,104 @@ def get(self) -> None:
 
 # -----------------------------------------------------------------------------
 
+class MetadataActionsBulkCreateHandler(BaseLTAHandler):
+    """Handler for /Metadata/actions/bulk_create."""
+
+    @lta_auth(roles=['admin', 'system'])
+    async def post(self) -> None:
+        """Handle POST /Metadata/actions/bulk_create."""
+        req = json_decode(self.request.body)
+        if 'bundle_uuid' not in req:
+            raise tornado.web.HTTPError(400, reason="missing bundle_uuid field")
+        if not isinstance(req['bundle_uuid'], str):
+            raise tornado.web.HTTPError(400, reason="bundle_uuid field is not a string")
+        if 'files' not in req:
+            raise tornado.web.HTTPError(400, reason="missing files field")
+        if not isinstance(req['files'], list):
+            raise tornado.web.HTTPError(400, reason="files field is not a list")
+        if not req['files']:
+            raise tornado.web.HTTPError(400, reason="files field is empty")
+
+        documents = []
+        bundle_uuid = req['bundle_uuid']
+        for file_catalog_uuid in req['files']:
+            documents.append({
+                "uuid": unique_id(),
+                "bundle_uuid": bundle_uuid,
+                "file_catalog_uuid": file_catalog_uuid,
+            })
+
+        logging.debug(f"MONGO-START: db.Metadata.insert_many(documents=[{len(documents)} documents])")
+        ret = await self.db.Metadata.insert_many(documents=documents)
+        logging.debug("MONGO-END:   db.Metadata.insert_many(documents)")
+        create_count = len(ret.inserted_ids)
+
+        uuids = []
+        for x in documents:
+            uuid = x["uuid"]
+            uuids.append(uuid)
+            logging.info(f"created Metadata {uuid}")
+
+        self.set_status(201)
+        self.write({'metadata': uuids, 'count': create_count})
+
+class MetadataActionsBulkDeleteHandler(BaseLTAHandler):
+    """Handler for /Metadata/actions/bulk_delete."""
+
+    @lta_auth(roles=['admin', 'system'])
+    async def post(self) -> None:
+        """Handle POST /Metadata/actions/bulk_delete."""
+        req = json_decode(self.request.body)
+        if 'metadata' not in req:
+            raise tornado.web.HTTPError(400, reason="missing metadata field")
+        if not isinstance(req['metadata'], list):
+            raise tornado.web.HTTPError(400, reason="metadata field is not a list")
+        if not req['metadata']:
+            raise tornado.web.HTTPError(400, reason="metadata field is empty")
```

Comment:
If you update to rest-tools v1.0.1, you can use the new argument getters for all this :smile:

Reply:
I just tried this but ran into a few issues. In theory:
```python
req = json_decode(self.request.body)
```
would translate most directly into:
```python
req = {
    "metadata": self.request("metadata", type_=list)
}
```
Writing code from scratch, one would probably just put the result in `metadata` and forego the struct, but that's neither here nor there.


The first check can be eliminated:
```python
        if 'metadata' not in req:
            raise tornado.web.HTTPError(400, reason="missing metadata field")
```
Because `get_argument` has `default: Any = arghandler.NO_DEFAULT,` to take care of raising on required fields that aren't supplied.


The second check must remain:
```python
        if not isinstance(req['metadata'], list):
            raise tornado.web.HTTPError(400, reason="metadata field is not a list")
```
Because `get_argument` does not do a type check, but rather does a type cast. This is convenient on valid inputs, but some invalid inputs can sneak through a type cast:
```
Python 3.7.2 (default, Feb  1 2019, 18:59:34) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> x = str([])
>>> x
'[]'
>>> y = list(x)
>>> y
['[', ']']
>>> 
```

Likewise the third check must remain:
```python
        if not req['metadata']:
            raise tornado.web.HTTPError(400, reason="metadata field is empty")
```
There is no provision in `get_argument` for raising on valid-but-empty arguments apart from `choices`. Unfortunately in this instance, we don't know in advance what the data will be, so `choices` can't help us here.

Reply:
See https://github.com/WIPACrepo/rest-tools/issues/12

Reply:
After my fix you should be able to do this:
`metadata = self.get_argument("metadata", type_=list)` or `metadata = self.get_body_argument("metadata", type_=list)`

Reply:
As for valid-but-empty argument values, this will still need to be done after the fact.

Reply:
well, I suppose I could add `invalid_choices`

Reply:
the paper trail continues: https://github.com/WIPACrepo/rest-tools/pull/13